[DLCov 2/5] Implement DebugLoc coverage tracking

SLTozer opened the pull request:
This is part of a series of patches that tries to improve DILocation bug detection in Debugify; see below for more details. This is the patch that adds the main feature, adding a set of `DebugLoc::get<Kind>` functions that can be used for instructions with intentionally empty DebugLocs to prevent Debugify from treating them as bugs, removing the currently-pervasive false positives and allowing us to use Debugify (in its original DI preservation mode) to reliably detect existing bugs and regressions. This patch does not add uses of these functions, except for once in Clang before optimizations, and in `Instruction::dropLocation()`, since that is an obvious case that immediately removes a set of false positives.

--------------------

This series of patches adds a "DebugLoc coverage tracking" feature, that inserts conditionally-compiled tracking information into DebugLocs (and by extension, to Instructions), which is used by Debugify to provide more accurate and detailed coverage reports. When enabled, this features tracks whether and why we have intentionally dropped a DebugLoc, allowing Debugify to ignore false positives. An optional additional feature allows also storing a stack trace of the point where a DebugLoc was unintentionally dropped/not generated, which is used to make fixing detected errors significantly easier. The goal of these features is to provide useful tools for developers to fix existing DebugLoc errors and allow reliable detection of regressions by either manual inspection or an automated script. 

Previous: https://github.com/llvm/llvm-project/pull/107278
Next: https://github.com/llvm/llvm-project/pull/107369

SLTozer committed:
Add conditionally-enabled DebugLocKinds

jryans commented on llvm/include/llvm/IR/DebugLoc.h:
```diff
@@ -22,6 +23,67 @@ namespace llvm {
+    Normal,
+    // DebugLoc intentionally does not have a valid DILocation; may be for a
+    // compiler-generated instruction, or an explicitly dropped location.
+    LineZero,
```
Hmm, I'd really prefer the name of this case to clearly state the "intentional" aspect. Perhaps `IntentionallyZero`...?

Another approach might be to have separate enum values for the different reasons, e.g. `CompilerGenerated`, `ExplicitlyDropped`, and whatever the others might be. Stating these cases very clearly, not just in comments, but also via enum value would appear to be a very valuable capturing of intent to me. (However, maybe it is too much trouble to do so...? Please say that if so.)

SLTozer replied:
Yeah, I didn't have a good name thought up for this - IntentionallyZero might be good, though even the word "Intentional" I'm not 100% on. Long run, the goal is to split this category into different subcategories as you suggested, but only at the point where there is some meaning to separating them - until then I felt it would be premature to try to assign categories, since it's not fully clear what they would be.

jryans replied:
Looking at it another way, I suppose my point is `LineZero` doesn't convey whether it's on purpose, an accident, or some undetermined mix of the two. But really, this case is trying to capture cases where explicit choices were made, so the name should reflect that somehow.

SLTozer replied:
Decided to tackle this by just splitting this straight into two categories - `CompilerGenerated` and `Dropped` - that each more specifically describe the reason for the empty DebugLoc. This more closely aligns with the original plans, which I'd veered off from because there wasn't any technical purpose to distinguishing the two cases, but I think having clear and meaningful names is enough reason!

jryans replied:
Thanks, this new version looks great to me! 😄

jryans commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
```diff
@@ -2080,6 +2081,10 @@ void DwarfDebug::beginInstruction(const MachineInstr *MI) {
   }
 
   if (!DL) {
+    // FIXME: We could assert that `DL.getKind() != DebugLocKind::Temporary`
+    // here, or otherwise record any temporary DebugLocs seen to ensure that
+    // transient compiler-generated instructions aren't leaking their DLs to
+    // other instructions.
```
Hmm yes, reporting such cases somehow does seem valuable... though unclear to me which way is best, so perhaps best left as future work for now.

jryans commented on llvm/lib/IR/DebugLoc.cpp:
```diff
@@ -11,6 +11,22 @@
+DebugLoc DebugLoc::getTemporary() { return DebugLoc(DebugLocKind::Temporary); }
+DebugLoc DebugLoc::getUnknown() { return DebugLoc(DebugLocKind::Unknown); }
+DebugLoc DebugLoc::getLineZero() { return DebugLoc(DebugLocKind::LineZero); }
+
+#else
+
```
I would maybe remove the blank lines around this `#else`, but perhaps that's just me.

jryans reviewed (commented):
Looks good overall, a few comments inline. 

Thanks for working on this feature, I believe this will be quite useful! 😄

jryans reviewed (approved):
This part looks good to me, thanks for working on this! 😄

SLTozer committed:
clang-format

SLTozer commented:
Rebased onto latest origin and added some documentation - quite confident now that this works, and hopefully some value has been demonstrated from the bugs found so far, so should be ready to land after the new changes (docs) are approved.

jryans reviewed (approved):
New docs look good, thanks for working on this!

Yes, the value is quite clear indeed. :smile:

jryans commented:
It would be great to also have some documentation that shows step by step how you've used these abilities to find the bugs you've found. Could be a separate PR with actual documentation, or something more informal stored externally... In any case, it would helpful to read about your bug-finding approach so that others can do the same (and also look for improvements as well).

SLTozer commented:
> It would be great to also have some documentation that shows step by step how you've used these abilities to find the bugs you've found.

SGTM - I think that should come in a separate patch. I have a branch with all my changes on it (not stable, I rebase semi-regularly) [here](https://github.com/SLTozer/llvm-project/tree/llvm-project-coverage-tracker), where the patches can be divided into 3 categories: bug fixes, the coverage tracking feature, and a variety of random tweaks and features I've added to display info and help myself triage bugs. I haven't yet put any work into getting the last category merge-ready, but I think that would be the appropriate place for such documentation to go.

SLTozer committed:
Fix underline length

SLTozer committed:
Fix bullet list formatting

SLTozer committed:
Some doc wording updates

SLTozer commented:
Just pinging codeowners: @dwblaikie @adrian-prantl @echristo does the documentation update look good? Since this is a form of Debug Info policy update (albeit not one that adds compulsory rules) ideally it'd have approval from a Debug Info maintainer, or at least a confirmed lack-of objections.

dwblaikie commented on llvm/docs/HowToUpdateDebugInfo.rst:
```diff
@@ -169,6 +169,47 @@ See the discussion in the section about
+  :ref:`dropping locations<WhenToDropLocation>`; this is set automatically by
+  ``Instruction::dropLocation()``.
+
+* ``DebugLoc::getUnknown()``: This indicates that the instruction does not have
```
Wonder whether it's worth splitting these cases from "there could be a right answer here, but it's Too Hard(tm) to find it right now" (places to look to fix bugs/improve location coverage) from "this is unrepresentable" (places that would benefit from representational changes, but aren't otherwise fixable)

But I don't feel strongly about it.

Otherwise the doc changes look good to me

SLTozer replied:
I originally considered a `DebugLoc::FixMe()` option for distinguishing those cases - in practise as I went around applying annotations, I found that in most cases where it wasn't clear what the right location was, it also wasn't clear what the solution would be or whether one would be possible without some broader change. At the moment, I _lean_ towards using a single "unknown" location, and if we find concrete unrepresentable cases then the next step would be to create a new special location kind for them; that way, we can also get a view as to what % of instructions end up without a location as a result of that particular limitation.

With that said, there is at least one clear set of cases right now that is just a "FixMe" (a lot of instructions produced by the vectorizers), so it would still be useful to have - I could go one way or the other.

dwblaikie replied:
Yeah, no worries - you've got th eboots on the ground experience, so go with your gut.

SLTozer committed:
Remove unnecessary config addition

SLTozer commented:
> Hi, I believe this merge is breaking external projects because you added `#include "llvm/Config/config.h"` which is not available (see comments at the top of that file).

Reverted for now, will look into relanding without the inclusion issue later.
